cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1226,27 +1238,52 @@ private class AccessPathNil extends AccessPath, TNil {
   override AccessPathFront getFront() {
     exists(DataFlowType t | this = TNil(t) | result = TFrontNil(t))
   }
+
+  override predicate pop(Content head, AccessPath tail) { none() }
 }
 
-private class AccessPathCons extends AccessPath, TCons {
+abstract private class AccessPathCons extends AccessPath { }
+
+private class AccessPathConsNil extends AccessPathCons, TConsNil {
   override string toString() {
-    exists(Content f, int len | this = TCons(f, len) |
-      result = f.toString() + ", ... (" + len.toString() + ")"
+    exists(Content f, DataFlowType t | this = TConsNil(f, t) |
+      result = f.toString() + " : " + ppReprType(t)
```

Comment:
Why print the type only for access paths of length 1? Also, could we add the suffix `" : " + ppReprType(t)` only when `ppReprType(t) != ""`? Some of the C# paths look weird, e.g. https://github.com/Semmle/ql/pull/1797/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.expected&file-filters%5B%5D=.qll#diff-7306039516f80cd888d60afc9fa1656fR3.

Reply:
We don't print the type for access paths > length 1 because we don't track their type. Before this PR, we didn't track the type for length = 1 either.

Reply:
Ah, I didn't notice that.

Reply:
Done.